File: src/main/java/net/rentalhost/plugins/php/hammer/services/OptionsPanelService.java
package net.rentalhost.plugins.php.hammer.services;

import net.miginfocom.swing.MigLayout;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

import javax.swing.*;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class OptionsPanelService {
    @NotNull
    private final JPanel optionsPanel;

    @NotNull
    public static JPanel create(@NotNull final Consumer<OptionsPanelService> delegateBuilder) {
        final OptionsPanelService component = new OptionsPanelService();

        delegateBuilder.accept(component);

        return component.optionsPanel;
    }

    private OptionsPanelService() {
        optionsPanel = new JPanel(new MigLayout("fillx"));
    }

    public void delegateRadioCreation(@NotNull final Consumer<RadioComponent> delegate) {
        final RadioComponent radioComponent = new RadioComponent();

        delegate.accept(radioComponent);

        if ((radioComponent.selectedOption == null) &&
            (radioComponent.radioOptions.size() >= 1)) {
            final RadioComponent.RadioOption firstWidget = radioComponent.radioOptions.get(0);
            firstWidget.radioButton.setSelected(true);
            firstWidget.updateConsumer.accept(true);
        }
    }

    public class RadioComponent {
        List<RadioOption> radioOptions = new ArrayList<>();
        private final ButtonGroup buttonGroup = new ButtonGroup();

        @Nullable
        private RadioOption selectedOption;

        public void addOption(
            @NotNull final String label,
            final boolean defaultValue,
            @NotNull final Consumer<Boolean> updateConsumer
        ) {
            final RadioOption newOption = new RadioOption(label, defaultValue, updateConsumer);
            radioOptions.add(newOption);

            if (defaultValue) {
                selectedOption = newOption;
            }
        }

        private class RadioOption {
            @NotNull final JRadioButton radioButton;

            @NotNull final Consumer<Boolean> updateConsumer;

            RadioOption(
                @NotNull final String label,
                final boolean defaultValue,
                @NotNull final Consumer<Boolean> updateConsumer
            ) {
                this.updateConsumer = updateConsumer;

                radioButton = new JRadioButton(label, defaultValue);
                radioButton.addItemListener((itemEvent) -> {
                    final boolean isSelected = radioButton.isSelected();

                    if ((selectedOption != null) &&
                        !Objects.equals(selectedOption, this)) {
                        selectedOption.updateConsumer.accept(false);
                    }

                    updateConsumer.accept(isSelected);
                    selectedOption = isSelected ? this : null;
                });

                optionsPanel.add(radioButton, "wrap");
                buttonGroup.add(radioButton);
            }
        }
    }
}

File: src/main/java/net/rentalhost/plugins/services/OptionsPanelService.kt
package net.rentalhost.plugins.services

import net.miginfocom.swing.MigLayout
import java.util.function.Consumer
import javax.swing.ButtonGroup
import javax.swing.JPanel
import javax.swing.JRadioButton

class OptionsPanelService private constructor() {
    private val optionsPanel: JPanel = JPanel(MigLayout("fillx"))

    fun delegateRadioCreation(delegate: Consumer<RadioComponent>) {
        val radioComponent = RadioComponent()

        delegate.accept(radioComponent)

        if (radioComponent.selectedOption == null && radioComponent.radioOptions.size >= 1) {
            val firstWidget = radioComponent.radioOptions[0]
            firstWidget.radioButton.isSelected = true
            firstWidget.updateConsumer.accept(true)
        }
    }

    inner class RadioComponent {
        internal val radioOptions: MutableList<RadioOption> = ArrayList()

        private val buttonGroup = ButtonGroup()

        internal var selectedOption: RadioOption? = null

        fun addOption(
            label: String,
            defaultValue: Boolean,
            updateConsumer: Consumer<Boolean>
        ) {
            val newOption = RadioOption(label, defaultValue, updateConsumer)

            radioOptions.add(newOption)

            if (defaultValue) {
                selectedOption = newOption
            }
        }

        inner class RadioOption internal constructor(
            label: String,
            defaultValue: Boolean,
            val updateConsumer: Consumer<Boolean>
        ) {
            val radioButton: JRadioButton

            init {
                radioButton = JRadioButton(label, defaultValue)
                radioButton.addItemListener {
                    val isSelected = radioButton.isSelected

                    if (selectedOption != null &&
                        selectedOption != this) {
                        selectedOption!!.updateConsumer.accept(false)
                    }

                    updateConsumer.accept(isSelected)

                    selectedOption = if (isSelected) this else null
                }

                optionsPanel.add(radioButton, "wrap")
                buttonGroup.add(radioButton)
            }
        }
    }

    companion object {
        fun create(delegateBuilder: Consumer<OptionsPanelService>): JPanel {
            val component = OptionsPanelService()

            delegateBuilder.accept(component)

            return component.optionsPanel
        }
    }
}
